### PR TITLE
Reduce allocations in Diagnostic.HasIntersectingLocation

### PR DESCRIPTION
**Pull request by sharwell**
Allocations were identified under scenario #26778.

* Avoid iterator allocations in Diagnostic.HasIntersectingLocation (1.9GB)
* Avoid boxing allocations in CSDiagnosticInfo.AdditionalLocations (1.4GB)

### CODE REVIEW

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs [hunk 1]
```diff
@@ -578,6 +565,112 @@ internal bool IsUnsuppressableError()
...
+        /// <summary>
+        /// Implements the following enumerator without allocations:
+        ///
+        /// <code>
+        /// if (this.Location.SourceTree == tree)
+        /// {
+        ///     yield return this.Location;
+        /// }
+        /// 
+        /// if (this.AdditionalLocations != null)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: 📝 It is rare for AdditionalLocations to be non-empty. An alternative to this pull request is to just add a fast-path to `HasIntersectingLocation` for the case where it is empty.

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs [hunk 2]
```diff
@@ -476,33 +476,45 @@ internal Diagnostic WithProgrammaticSuppression(ProgrammaticSuppressionInfo prog
...
-            if (this.AdditionalLocations != null)
+            static IEnumerable<Location> getDiagnosticLocationsWithinTree(Location location, IReadOnlyList<Location> additionalLocations, SyntaxTree tree)
             {
-                foreach (var additionalLocation in this.AdditionalLocations)
+                if (location.SourceTree == tree)
+                {
+                    yield return location;
+                }
+
+                if (additionalLocations is object && additionalLocations.Count > 0)
```
Comment (cston) [inline src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: >if (additionalLocations is object && additionalLocations.Count > 0) [](start = 16, length = 67)

These cases are already handled in the caller.

File: src/Compilers/CSharp/Portable/Errors/CSDiagnosticInfo.cs [hunk 3]
```diff
@@ -34,7 +35,7 @@ internal CSDiagnosticInfo(ErrorCode code, object[] args, ImmutableArray<Symbol>
         {
             // Internal errors are abnormal and should not occur except where there are bugs in the compiler.
             Debug.Assert(code != ErrorCode.ERR_InternalError);
-            _additionalLocations = additionalLocations;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Errors/CSDiagnosticInfo.cs]: Shouldn't we be changing the type of the `additionalLocations` parameter to be `IReadOnlyList<Location>` as well? Leaving it as `ImmutableArray<Location>` means callers could be unnecessarily allocating an `ImmutableArray<Location>`.

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: I'll change it to an assertion.

Comment (jaredpar) [inline src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: Isn't this entire `if` statement unnecessary given it's checked at the start of the containing method?

File: src/Compilers/CSharp/Portable/Errors/CSDiagnosticInfo.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Errors/CSDiagnosticInfo.cs]: In theory, maybe. In practice I did not observe this.

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: ➡️ Covered by https://github.com/dotnet/roslyn/pull/40695#discussion_r362527239 (will remove)

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs [hunk 4]
```diff
@@ -476,29 +476,38 @@ internal Diagnostic WithProgrammaticSuppression(ProgrammaticSuppressionInfo prog
...
+                return !filterSpan.HasValue || filterSpan.GetValueOrDefault().IntersectsWith(location.SourceSpan);
             }
 
-            if (this.AdditionalLocations != null)
+            static IEnumerable<Location> getDiagnosticLocationsWithinTree(Location location, IReadOnlyList<Location> additionalLocations, SyntaxTree tree)
             {
-                foreach (var additionalLocation in this.AdditionalLocations)
+                if (location.SourceTree == tree)
+                {
+                    yield return location;
```
Comment (cston) [inline src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: >yield return location; [](start = 20, length = 22)

Consider removing the `location` parameter from this local function and always check `Location` in the caller.

Comment (sharwell) [reply src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: Yeah I like that better. Good suggestion. 👍

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs [hunk 5]
```diff
@@ -476,35 +476,36 @@ internal Diagnostic WithProgrammaticSuppression(ProgrammaticSuppressionInfo prog
...
                 }
             }
 
             return false;
-        }
 
-        private IEnumerable<Location> GetDiagnosticLocationsWithinTree(SyntaxTree tree)
-        {
-            if (this.Location.SourceTree == tree)
+            // Local functions
```
Comment (cston) [inline src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: >// Local functions [](start = 12, length = 18)

Consider removing the comment.

File: src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs [hunk 6]
```diff
@@ -476,35 +476,36 @@ internal Diagnostic WithProgrammaticSuppression(ProgrammaticSuppressionInfo prog
         /// </summary>
         internal bool HasIntersectingLocation(SyntaxTree tree, TextSpan? filterSpanWithinTree = null)
```
Comment (333fred) [inline src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: Should we have a `PerformanceSensitive` attribute on the method?

Comment (sharwell) [reply src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs]: In the original change, adding the attribute was not easy. In the form we finally arrived at, it seems to make sense. I have now document both performance constraints.

**Pull request merged**